### DISCUSSION TITLE
Query analytics

### DISCUSSION THREAD

**Discussion by rrva** [graphql-java/graphql-java#3352] [Q&A]
Currently we implement query analytics in this way:

We mirror some of the traffic to query analytics instances. These instances run just a small sample of the incoming traffic and using a instrumentation (which we can disable on the non-mirror leg) they measure which fields are being requested. Is there another way to do this? We have telemetry which captures the queries and variables, so potentially we could do some offline analysis instead? How do folks do query analytics?

**Comment by bbakerman** [comment]
At Atlassian where I work we use Apollo Studio for analytics.  We capture the query details in an off line thread (so we dont make the query wait) and we sent them into this off premise cloud service.

The format is a Protobuf file and Apollo do not publish a JVM agent for this.  We reverse engineered this Protobuf which is a PAIN in a support sense.

Apollo works really well for seeing what fields are used and so on.  But its expensive.  You pay by traffic (or at least do)

We looked at using our Telemetry system (SignalFX) for capturing per field stats BUT the problem is cardinality.  Most observability systems work on the idea that there limited cardinality to a "metric".  Our schema has 8000 fields and hence there is a cardinality explosion that most time series database canr handle (or can't handle for a reasonable cost)

We have looked at alternatives of build your own but the database for us is the sticking point - we are receiving 15K requests per second and if we store all of these we have a size challenge but it seems its not size per se but cardinality is our challenge.

**Comment by rrva** [comment]
Previously we mirrored the incoming HTTP traffic in the load balancers to special query stats nodes which ran the query with a stats recording instrumentation. But now we switched our approach; we send the query on a kafka topic and the query stats node runs

```
fun analyzeAndRecord(query: String, variables: Map<String, Any>?, userAgent: String) {
        if (query.contains("__schema")) {
            return
        }
        val queryDoc = parser.parseDocument(query)
        val coercedVariables = variables?.let { CoercedVariables(it) } ?: CoercedVariables.emptyVariables()
        val normalizedQuery = ExecutableNormalizedOperationFactory.createExecutableNormalizedOperation(
            schema,
            queryDoc,
            "",
            coercedVariables
        )
        normalizedQuery.topLevelFields.forEach {
            recordNormalizedField(it, userAgent)
        }
    }
```

and then we base the query statistics on that. This is significantly faster.
